src/main/scala/mesosphere/marathon/core/check/Check.scala

```diff
@@ -0,0 +1,245 @@
+package mesosphere.marathon
+package core.check
+
+import com.wix.accord._
+
+import mesosphere.marathon.Protos.CheckDefinition.Protocol
+import mesosphere.marathon.state._
+import org.apache.mesos.{Protos => MesosProtos}
+
+import scala.concurrent.duration._
+
+sealed trait Check {
+  def delay: FiniteDuration
+  def interval: FiniteDuration
+  def timeout: FiniteDuration
+  def toProto: Protos.CheckDefinition
+
+  protected def protoBuilder: Protos.CheckDefinition.Builder =
+    Protos.CheckDefinition.newBuilder
+      .setIntervalSeconds(this.interval.toSeconds.toInt)
+      .setTimeoutSeconds(this.timeout.toSeconds.toInt)
+      .setDelaySeconds(this.delay.toSeconds.toInt)
+}
+
+sealed trait PortReference extends Product with Serializable {
+  def apply(assignments: Seq[PortAssignment]): PortAssignment
+  def buildProto(builder: Protos.CheckDefinition.Builder): Unit
+}
+
+object PortReference {
+  case class ByIndex(value: Int) extends PortReference {
+    override def apply(assignments: Seq[PortAssignment]): PortAssignment =
+      assignments(value)
+    override def buildProto(builder: Protos.CheckDefinition.Builder): Unit =
+      builder.setPortIndex(value)
+  }
+
+  def apply(value: Int): PortReference = ByIndex(value)
+
+  def fromProto(pb: Protos.CheckDefinition): Option[PortReference] =
+    if (pb.hasPortIndex) Some(ByIndex(pb.getPortIndex))
+    else if (!pb.hasPort) Some(ByIndex(0)) // backward compatibility, this used to be the default value in marathon.proto
+    else None
+}
+
+sealed trait CheckWithPort extends Check {
+  def portIndex: Option[PortReference]
+  def port: Option[Int]
+}
+
+object CheckWithPort {
+  val DefaultPortIndex = None
+  val DefaultPort = None
+
+  import mesosphere.marathon.api.v2.Validation.isTrue
+
+  implicit val Validator: Validator[CheckWithPort] =
+    isTrue("Check must specify either a port or a portIndex") { hc =>
+      hc.portIndex.isDefined ^ hc.port.isDefined
+    }
+}
+
+sealed trait MesosCheck extends Check {
+  def interval: FiniteDuration
+  def timeout: FiniteDuration
+  def delay: FiniteDuration
+
+  override protected def protoBuilder: Protos.CheckDefinition.Builder =
+    super.protoBuilder.setDelaySeconds(delay.toSeconds.toInt)
+
+  def toMesos(portAssignments: Seq[PortAssignment] = Seq.empty): Option[MesosProtos.CheckInfo]
+}
+
+sealed trait MesosCheckWithPorts extends CheckWithPort { this: Check =>
+  def effectivePort(portAssignments: Seq[PortAssignment]): Option[Int] = {
```

Comment:
Can we please move this to a companion object that accepts port as a parameter? Preference to avoid traits with default implementations that depend on abstract members.

Reply:
@timcharper  this is consistent with health checks... it was intended to be consistent with health checks as well.   Should we refactor HCs as well?

Reply:
I would be in favor of it :-)

Reply:
But, at worst, I would prefer to not copy a bad pattern since this is not public API

Reply:
https://jira.mesosphere.com/browse/MARATHON-8595

Reply:
added a tech debt jira for this... I totally agree with you but also expect refactoring consequences to ripple thru